avalon/nuke/lib.py

```diff
@@ -3,82 +3,283 @@
+         "type": "Text_Knob"},
+        {"name": '__divider__'},
+        {"name": 'begin',
+         "value": 'Avalon data group',
+         "type": "Tab_Knob", "group": -1}
+    ]
+    non_hiden = ["asset", "subset", "name", "namespace"]
+
     try:
-        node['avalon'].value()
-    except Exception:
-        tab = nuke.Tab_Knob("Avalon")
-        uk = nuke.Text_Knob('avalon', 'avalon data')
-        node.addKnob(tab)
-        node.addKnob(uk)
-        log.info("created new user knob avalon")
+        # create Avalon Tab and basic knobs
+        for k in knobs[:-1]:
+            if not k.get("group"):
+                if "__divider__" in k["name"]:
+                    knob = nuke.Text_Knob("")
+                    node.addKnob(knob)
+                elif k["name"] not in node.knobs().keys():
+                    knob = eval("nuke.{type}('{name}')".format(**k))
+                    node.addKnob(knob)
+                    try:
+                        knob.setValue(k['value'])
+                    except TypeError as E:
+                        print(E)
+            else:
+                if k["name"] not in node.knobs().keys():
+                    knob = eval((
+                        "nuke.{type}('{name}',"
+                        "'{value}', {group})"
+                    ).format(**k))
+                    node.addKnob(knob)
+
+        # add avalon knobs for imprinting data
+        for k, v in data.items():
+            label = k
+            name = prefix + label
+            value = str(v)
+            if label in non_hiden:
+                if name not in node.knobs().keys():
+                    log.info("Setting: `{0}` to `{1}`".format(name, value))
+                    knob = eval((
+                        "nuke.String_Knob('{name}','{label}',"
+                        "'{value}')"
+                    ).format(
+                        name=name,
+                        label=label,
+                        value=value
+                    ))
+                    node.addKnob(knob)
+                else:
+                    log.info("Updating: `{0}` to `{1}`".format(name, value))
+                    node[name].setValue(value)
+            else:
+                if name not in node.knobs().keys():
+                    log.info("Setting: `{0}` to `{1}`".format(name, value))
+                    knob = eval((
+                        "nuke.Text_Knob('{name}','{label}',"
+                        "'{value}')"
+                    ).format(
+                        name=name,
+                        label=label,
+                        value=value
+                    ))
+                    node.addKnob(knob)
+                else:
+                    log.info("Updating: `{0}` to `{1}`".format(name, value))
+                    node[name].setValue(str(value))
+
+        # adding closing group knob
+        knob = eval(
+            "nuke.{type}('{name}', '{value}', {group})".format(**knobs[-1]))
+        node.addKnob(knob)
```

Comment:
Could you elaborate why this closing group knob is being done separately? Doesn't this mean it'll try to add this closing tab every time, even when it already existed?